plugins/inputs/openstack/openstack.go

```diff
@@ -0,0 +1,908 @@
+	"github.com/gophercloud/gophercloud/openstack/compute/v2/extensions/hypervisors"
+	nova_services "github.com/gophercloud/gophercloud/openstack/compute/v2/extensions/services"
+	"github.com/gophercloud/gophercloud/openstack/compute/v2/flavors"
+	"github.com/gophercloud/gophercloud/openstack/compute/v2/servers"
+	"github.com/gophercloud/gophercloud/openstack/identity/v3/projects"
+	"github.com/gophercloud/gophercloud/openstack/identity/v3/services"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/extensions/agents"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/networks"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/ports"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/subnets"
+	"github.com/gophercloud/gophercloud/openstack/orchestration/v1/stacks"
+	"github.com/influxdata/telegraf"
+	httpconfig "github.com/influxdata/telegraf/plugins/common/http"
+	"github.com/influxdata/telegraf/plugins/inputs"
+)
+
+const (
+	// plugin is used to identify ourselves in log output
+	plugin = "openstack"
+)
+
+var (
+	type_port    = regexp.MustCompile(`_rx$|_rx_drop$|_rx_errors$|_rx_packets$|_tx$|_tx_drop$|_tx_errors$|_tx_packets$`)
+	type_cpu     = regexp.MustCompile(`cpu[0-9]{1,2}_time$`)
+	type_storage = regexp.MustCompile(`_errors$|_read$|_read_req$|_write$|_write_req$`)
+)
+
+// volume is a structure used to unmarshal raw JSON from the API into.
+type volume struct {
+	volumes.Volume
+	volumetenants.VolumeTenantExt
+}
+
+// OpenStack is the main structure associated with a collection instance.
+type OpenStack struct {
+	// Configuration variables
+	IdentityEndpoint string          `toml:"authentication_endpoint"`
+	Domain           string          `toml:"domain"`
+	Project          string          `toml:"project"`
+	Username         string          `toml:"username"`
+	Password         string          `toml:"password"`
+	EnabledServices  []string        `toml:"enabled_services"`
+	ServerDiagnotics bool            `toml:"server_diagnotics"`
+	OutputSecrets    bool            `toml:"output_secrets"`
+	Log              telegraf.Logger `toml:"-"`
+	httpconfig.HTTPClientConfig
+
+	// Locally cached clients
+	identity *gophercloud.ServiceClient
+	compute  *gophercloud.ServiceClient
+	volume   *gophercloud.ServiceClient
+	network  *gophercloud.ServiceClient
+	stack    *gophercloud.ServiceClient
+
+	// Locally cached resources
+	flavors     map[string]flavors.Flavor
+	hypervisors []hypervisors.Hypervisor
+	diag        map[string]interface{}
+	projects    map[string]projects.Project
+	services    map[string]services.Service
+}
+
+// ContainsService indicates whether a particular service is enabled
+func (o *OpenStack) ContainsService(t string) bool {
+	for _, service := range o.services {
+		if service.Type == t {
+			return true
+		}
+	}
+
+	return false
+}
+
+// InEnabledServices indicates whether a particular service is included in EnabledServices
+func (o *OpenStack) InEnabledServices(t string) bool {
```

Comment:
Same here, please do only export functions that need to be called from outside of the plugin. This usually boils down to the interface-functions required by Telegraf.